Update README.md

briansoe66 opened the pull request:
Clarify what QoS settings can be configured via the XML. 
This addresses #500

fujitatomoya reviewed (approved):
LGTM with DCO

JLBuenoLopez-eProsima reviewed (commented):
I would suggest clarifying that the environment variable `RMW_FASTRTPS_USE_QOS_FROM_XML` only applies to override two specific QoS settings that are, by default, configured in the RMW layer. Those two QoS are the History Memory Policy and the Publication Mode. If you want to change the default behaviour of `rmw_fastrtps` you should set the environment variable.
However, there are other QoS not set in the `rclcpp`/`rclpy` layer that can be configured using the XML file without setting the environment variable `RMW_FASTRTPS_USE_QOS_FROM_XML`.

EduPonz commented on README.md:
```diff
@@ -60,13 +60,11 @@ If `RMW_FASTRTPS_PUBLICATION_MODE` is not set, then both `rmw_fastrtps_cpp` and
-This means that setting any of them in the XML files has no effect, since they do not override what was used to create the publisher, subscription, service, or client.
-1. In order to modify the history memory policy or publication mode using XML files, environment variable `RMW_FASTRTPS_USE_QOS_FROM_XML` must be set to 1 (it is set to 0 by default).
-This tells `rmw_fastrtps` that it should override both the history memory policy and the publication mode using the XML.
+1. QoS configuration contained in `rclcpp`/`rclpy` QoS class, can only be set using `rclcpp`/`rclpy`.
```
I'd add a link for reference here

EduPonz commented on README.md:
```diff
@@ -60,13 +60,11 @@ If `RMW_FASTRTPS_PUBLICATION_MODE` is not set, then both `rmw_fastrtps_cpp` and
-This tells `rmw_fastrtps` that it should override both the history memory policy and the publication mode using the XML.
+1. QoS configuration contained in `rclcpp`/`rclpy` QoS class, can only be set using `rclcpp`/`rclpy`.
+Setting any of them in the XML files has no effect, since they do not override what was used to create the publisher, subscription, service, or client.
+1. QoS configuration not contained `rclcpp`/`rclpy`, such as the history memory policy or publication mode, can be overridden using XML files. The environment variable `RMW_FASTRTPS_USE_QOS_FROM_XML` must be set to 1 (it is set to 0 by default).
```
Maybe a link to Fast DDS XML config documentation page would be good here

briansoe66 replied:
There is a link to the Fast DDS XML doc on line 63

hidmic commented on README.md:
```diff
@@ -60,13 +60,11 @@ If `RMW_FASTRTPS_PUBLICATION_MODE` is not set, then both `rmw_fastrtps_cpp` and
 
-With `rmw_fastrtps`, it is possible to fully configure Fast DDS using an XML file as described in [Fast DDS documentation](https://fast-dds.docs.eprosima.com/en/latest/fastdds/xml_configuration/xml_configuration.html).
-When configuring the middleware using XML files, there are certain points that have to be taken into account:
+It is possible to fully configure Fast DDS through a combination of 'rclcpp'/'rclpy' QoS, and the XML file as described in [Fast DDS documentation](https://fast-dds.docs.eprosima.com/en/latest/fastdds/xml_configuration/xml_configuration.html).
```
@briansoe66 here and below, I'd rather not mention an specific client library. There are others besides the core ones.

hidmic commented on README.md:
```diff
@@ -60,13 +60,11 @@ If `RMW_FASTRTPS_PUBLICATION_MODE` is not set, then both `rmw_fastrtps_cpp` and
-This means that setting any of them in the XML files has no effect, since they do not override what was used to create the publisher, subscription, service, or client.
-1. In order to modify the history memory policy or publication mode using XML files, environment variable `RMW_FASTRTPS_USE_QOS_FROM_XML` must be set to 1 (it is set to 0 by default).
-This tells `rmw_fastrtps` that it should override both the history memory policy and the publication mode using the XML.
+1. QoS configuration contained in `rclcpp`/`rclpy` QoS class, can only be set using `rclcpp`/`rclpy`.
```
@briansoe66 is this true? What if QoS policies are all set to `SYSTEM_DEFAULT` values?

EduPonz replied:
You're actually correct @hidmic . [Here](https://github.com/ros2/rmw_fastrtps/blob/47be8ba4ed14303659e29094e72033b56449c15a/rmw_fastrtps_shared_cpp/src/qos.cpp#L55), you can see that in the case of `SYSTEM_DEFAULT`, the QoS are not overriden

hidmic replied:
Awesome. @briansoe66 would you mind updating this patch to reflect that?

hidmic commented:
Besides the failing DCO check, this patch still needs update @briansoe66.

briansoe66 commented:
@hidmic , I am waiting for answers from @EduPonz to my three questions.

EduPonz commented:
Hi @briansoe66 , I'm afraid I must have missed them. Please, can you clarify which questions are you referring to?

briansoe66 commented:
@EduPonz , you stated:
> You're actually correct @hidmic . Here, you can see that in the case of SYSTEM_DEFAULT, the QoS are not overriden

Could you clarify:
1. How can I set the QoS policy to use the SYSTEM_DEFAULT values?
2. In the case of SYSTEM_DEFAULT, does it use the default settings for Fast-RTPS? Like the default is up to the client implementation, and is not standardized by rclcpp / rclpy?
3. If so, doesn't that contradict the original readme that "QoS set by rclcpp/rclpy are always honored" (line 66)?

EduPonz commented:
Hi @briansoe66 ,

1. You can set the QoS to SYSTEM_DEFAULT using the `rclcpp`/`rclpy` API when creating the publisher/subscription. You can read more about that in:
    1. [ROS 2 QoS docs](https://index.ros.org/doc/ros2/Concepts/About-Quality-of-Service-Settings/)
    2. [RMW API reference](http://docs.ros2.org/latest/api/rmw/structrmw__qos__profile__t.html)
1. This is a default specified by ROS 2 (it's specified in the first link). Not explicitly modifying a field of `rmw_qos_profile_t` will result in using the ROS 2 default value. However, the user can let the implementation decide, which is what SYSTEM_DEFAULT means in this case.
1. You are right, the README was not correct. I corrected it yesterday in #504 . Now, it states:
```
ROS 2 QoS contained in rmw_qos_profile_t are always honored, unless set to *_SYSTEM_DEFAULT. In that case, XML values, or Fast DDS default values in the absences of XML ones, are applied. Setting any QoS in rmw_qos_profile_t to something other than *_SYSTEM_DEFAULT entails that specifying it via XML files has no effect, since they do not override what was used to create the publisher, subscription, service, or client.
```

I hope that soles your questions. Just contact me if I can provide further help. Thanks for the iterations and the patience!

hidmic commented on README.md:
```diff
@@ -60,27 +60,41 @@ If `RMW_FASTRTPS_PUBLICATION_MODE` is not set, then both `rmw_fastrtps_cpp` and
+auto rcl_publisher = my_rclcpp_publisher->get_publisher_handle();
+const auto rmw_publisher = rcl_publisher_get_rmw_handle(rcl_publisher);
+rmw_qos_profile_t qos;
+rmw_publisher_get_actual_qos(rmw_publisher, &qos);
```
@briansoe66 consider restricting the example to the `rmw` layer.

hidmic replied:
This still applies !

hidmic commented on README.md:
```diff
@@ -60,27 +60,41 @@ If `RMW_FASTRTPS_PUBLICATION_MODE` is not set, then both `rmw_fastrtps_cpp` and
 
-With `rmw_fastrtps`, it is possible to fully configure Fast DDS using an XML file as described in [Fast DDS documentation](https://fast-dds.docs.eprosima.com/en/latest/fastdds/xml_configuration/xml_configuration.html).
-When configuring the middleware using XML files, there are certain points that have to be taken into account:
+Fast DDS QoS policies can be fully configured through a combination of the [`rclcpp`/`rclpy` QoS] API, and the [Fast DDS XML] file's QoS elements. Configuraton depends on the environment variable `RMW_FASTRTPS_USE_QOS_FROM_XML`.
```
@briansoe66 as I mentioned before, I'd rather not target specific client libraries. You can always link against the [`rmw`](https://github.com/ros2/rmw) interface layer for API documentation.

hidmic commented on README.md:
```diff
@@ -60,27 +60,41 @@ If `RMW_FASTRTPS_PUBLICATION_MODE` is not set, then both `rmw_fastrtps_cpp` and
+[publication mode]: https://fast-dds.docs.eprosima.com/en/latest/fastdds/dds_layer/core/policy/eprosimaExtensions.html#publishmodeqospolicy
 
-1. Setting `RMW_FASTRTPS_USE_QOS_FROM_XML` effectively overrides whatever configuration was set with `RMW_FASTRTPS_PUBLICATION_MODE`, setting the publication mode to Fast DDS' default publication mode unless specified otherwise using an XML file.
+Note: If `RMW_FASTRTPS_USE_QOS_FROM_XML` is set to 1, and [history memory policy] or [publication mode] are not specified in the XML, then the Fast DDS' defaults configurations will be used: 
```
@briansoe66 nit: 
```suggestion
Note: If `RMW_FASTRTPS_USE_QOS_FROM_XML` is set to 1, and [history memory policy] or [publication mode] are not specified in the XML, then the Fast DDS' default configurations will be used: 
```

hidmic commented on README.md:
```diff
@@ -60,27 +60,41 @@ If `RMW_FASTRTPS_PUBLICATION_MODE` is not set, then both `rmw_fastrtps_cpp` and
+[history memory policy]: https://fast-dds.docs.eprosima.com/en/latest/fastdds/dds_layer/core/policy/eprosimaExtensions.html#rtpsendpointqos
+[publication mode]: https://fast-dds.docs.eprosima.com/en/latest/fastdds/dds_layer/core/policy/eprosimaExtensions.html#publishmodeqospolicy
 
-1. Setting `RMW_FASTRTPS_USE_QOS_FROM_XML` effectively overrides whatever configuration was set with `RMW_FASTRTPS_PUBLICATION_MODE`, setting the publication mode to Fast DDS' default publication mode unless specified otherwise using an XML file.
```
@briansoe66 why remove this sentence?

hidmic replied:
This still applies !

hidmic reviewed (commented):
@briansoe66 I'd advise against referring to `rclcpp` directly (it's not the only client library, there are many more). Consider targeting [`rmw`'s interface](https://github.com/ros2/rmw) for API demonstration.

briansoe66 commented:
@hidmic , I addressed your comments. Something is weird with the diffs though, not sure how to fix that.

I created this PR to clarify the nuances of setting QoS, by explicitly stating all different methods. But I cannot dedicate more time to this. You are welcome to take over the changes, or we can drop it.

EduPonz commented:
Hi @briansoe66 @hidmic ,

I can take over this if that's ok with you, although probably not today.

briansoe66 commented:
@EduPonz , that would be great, thank you!

EduPonz commented:
I do not have privileges to push to @briansoe66 fork, so I've taken the commits onto a branch `eProsima:feature/briansoe66/readme` in eProsima's fork. I hope that's OK with everyone. This PR can be closed when merging #506

clalancette commented:
Since #506 has been merged, I'm going to close this out.
